File: src/platform/code_intel/routers/deps.py
from __future__ import annotations
from pathlib import Path
from fastapi import APIRouter
from code_intel.config import RepoPaths
from code_intel.registry import registry

router = APIRouter(prefix="/repos/{repo_id}/deps", tags=["deps"])

def _paths(repo_id: str, data_dir: str) -> RepoPaths:
    return RepoPaths(Path(data_dir), repo_id)

@router.get("/graph")
async def get_import_graph(repo_id: str, language: str | None = None, min_imports: int = 1, data_dir: str = "data"):
    paths = _paths(repo_id, data_dir)
    api = registry.get_dep_api()
    return api.get_import_graph(paths.db_path, language=language, min_imports=min_imports)

@router.get("/files/{path:path}")
async def get_file_imports(repo_id: str, path: str, data_dir: str = "data"):
    paths = _paths(repo_id, data_dir)
    api = registry.get_dep_api()
    return api.get_file_imports(paths.db_path, path)

@router.get("/circular")
async def get_circular_deps(repo_id: str, data_dir: str = "data"):
    paths = _paths(repo_id, data_dir)
    api = registry.get_dep_api()
    return api.get_circular_deps(paths.db_path)

@router.get("/external")
async def get_external_packages(repo_id: str, data_dir: str = "data"):
    paths = _paths(repo_id, data_dir)
    api = registry.get_dep_api()
    return api.get_external_packages(paths.db_path)

@router.get("/stats")
async def get_dependency_stats(repo_id: str, data_dir: str = "data"):
    paths = _paths(repo_id, data_dir)
    api = registry.get_dep_api()
    return api.get_dependency_stats(paths.db_path)

File: src/frontend/src/api/deps.ts
import { client } from './client';

// === Types ===

export interface ImportInfo {
    import_id: number;
    src_entity_id: number;
    dst_entity_id: number;
    src_path: string;
    dst_path: string;
    import_type: 'static' | 'dynamic';
    is_external: boolean;
    is_dynamic: boolean;
}

export interface CircularDep {
    cycle_id: number;
    cycle_path: string[];
    cycle_length: number;
}

export interface ExternalPackage {
    name: string;
    usage_count: number;
    importing_files: number;
}

export interface ImportGraphNode {
    id: number;
    path: string;
    name: string;
    kind: string;
    language?: string;
    is_external: boolean;
}

export interface ImportGraphEdge {
    source: number;
    target: number;
    import_type: string;
    is_dynamic: boolean;
}

export interface ImportGraph {
    nodes: ImportGraphNode[];
    edges: ImportGraphEdge[];
}

export interface FileImports {
    file_path: string;
    imports: Array<{
        path: string;
        import_type: string;
        is_external: boolean;
    }>;
    imported_by: Array<{
        path: string;
        import_type: string;
    }>;
}

export interface DepsStats {
    total_imports: number;
    total_files: number;
    external_packages: number;
    circular_dependencies: number;
    avg_imports_per_file: number;
    max_fan_out: number;
    max_fan_in: number;
}

// === API Functions ===

export const getImportGraph = (repoId: string, filters?: {
    language?: string;
    includeExternal?: boolean;
    minImports?: number;
}) =>
    client.get<ImportGraph>(`/repos/${repoId}/deps/graph`, { params: filters }).then(res => res.data);

export const getFileImports = (repoId: string, filePath: string) =>
    client.get<FileImports>(`/repos/${repoId}/deps/files/${encodeURIComponent(filePath)}/imports`).then(res => res.data);

export const getCircularDeps = (repoId: string) =>
    client.get<CircularDep[]>(`/repos/${repoId}/deps/circular`).then(res => res.data);

export const getExternalPackages = (repoId: string) =>
    client.get<ExternalPackage[]>(`/repos/${repoId}/deps/external`).then(res => res.data);

export const getDepsStats = (repoId: string) =>
    client.get<DepsStats>(`/repos/${repoId}/deps/stats`).then(res => res.data);
